Added check-labels munger and corresponding tests.

foxish opened the pull request:
https://github.com/kubernetes/contrib/issues/1423
The munger checks for changes in label config yaml file and creates labels as necessary in the repository.

foxish committed:
Added check-labels munger and corresponding tests.

The munger checks for
changes in label config yaml file and creates labels as necessary in the repository.

foxish commented:
@janetkuo @lavalamp Please review.

janetkuo commented on mungegithub/mungers/check-labels_test.go:
```diff
@@ -0,0 +1,118 @@
+			name:       "Adding new labels with existing labels in repository.",
+			repoLabels: repoLabels,
+			expected:   fileLabels[0:2],
+		},
```
What if fileLabels contains a label with the same name but different color?

foxish replied:
As far as I checked, the label name needs to be unique, or github won't let us add it, even with a different color. In check-labels.go:103, the name of the label is checked for uniqueness and we don't add it in the above case.

janetkuo replied:
Yes add a test case for that sounds good.

foxish replied:
Done.

eparis commented on mungegithub/github/github.go:
```diff
@@ -1692,3 +1696,65 @@ func (config *Config) ListAllIssues(listOpts *github.IssueListByRepoOptions) ([]
+}
+
+// AddLabel adds a single github label to the repository.
+func (config *Config) AddLabel(label *github.Label) error {
```
I much prefer the analytics be collected and we print a log message for any 'write' type operation before the `DryRun` return.

foxish replied:
Done.

eparis commented on mungegithub/mungers/check-labels.go:
```diff
@@ -0,0 +1,140 @@
+
+// EachLoop is called at the start of every munge loop
+func (c *CheckLabelsMunger) EachLoop() error {
+	fileContents, sha, err := c.labelAccessor.GetFileContents(c.labelFileName)
```
We already have the whole master git tree which is used by the blunderbuss assignment. Do we need to use api calls to fetch this data via the github api?

foxish replied:
@eparis Since we are making mungegithub generic in order to work with other repositories as well, would it be fair to assume that all of them will have the same fetching of the master git tree? I designed it to use the API because then it is independent of whether the blunderbuss assigner is being run. However, if we will be running blunderbuss against other repositories in the same way, I'll go ahead and make the change to read the local file from the pull for blunderbuss.

lavalamp replied:
Yes, that would be the idea-- we'd want all our repositories to be
consistent.

On Wed, Jul 6, 2016 at 3:44 PM, Anirudh Ramanathan <notifications@github.com
> wrote:

> In mungegithub/mungers/check-labels.go
> <https://github.com/kubernetes/contrib/pull/1306#discussion_r69825158>:
>
> > +// RequiredFeatures is a slice of 'features' that must be provided.
> > +func (c *CheckLabelsMunger) RequiredFeatures() []string { return []string{} }
> > +
> > +// Initialize will initialize the munger.
> > +func (c *CheckLabelsMunger) Initialize(config *githubhelper.Config, features *features.Features) error {
> > +	file := c.labelFileName
> > +	if len(file) == 0 {
> > +		glog.Fatalf("No --label-file= supplied, cannot check labels")
> > +	}
> > +	c.labelAccessor = config
> > +	return nil
> > +}
> > +
> > +// EachLoop is called at the start of every munge loop
> > +func (c *CheckLabelsMunger) EachLoop() error {
> > +	fileContents, sha, err := c.labelAccessor.GetFileContents(c.labelFileName)
>
> @eparis <https://github.com/eparis> Since we are making mungegithub
> generic in order to work with other repositories as well, would it be fair
> to assume that all of them will have the same fetching of the master git
> tree? I designed it to use the API because then it is independent of
> whether the blunderbuss assigner is being run. However, if we will be
> running blunderbuss against other repositories in the same way, I'll go
> ahead and make the change to read the local file from the pull for
> blunderbuss.
>
> —
> You are receiving this because you were assigned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/contrib/pull/1306/files/d6ff9bf9494587441eb06d2c7fbc69e7b9739b97#r69825158>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/AAnglrvMDxGgX0pereE6K53SxrRVEKxeks5qTC_egaJpZM4JDbTs>
> .
>

eparis commented on mungegithub/mungers/check-labels.go:
```diff
@@ -0,0 +1,140 @@
+		glog.Errorf("Failed to read the check label config: %v", err)
+		return err
+	}
+	if c.lastSha != sha {
```
This gets run every couple of minutes. Is the extra code to not process the file really worth it?

foxish replied:
When we process the file, we also pull all the labels from the repository, which needs multiple API calls. It is avoided when we see that the labels config file has not changed.

eparis commented on mungegithub/mungers/check-labels.go:
```diff
@@ -0,0 +1,140 @@
+
+// AddFlags will add any request flags to the cobra `cmd`
+func (c *CheckLabelsMunger) AddFlags(cmd *cobra.Command, config *githubhelper.Config) {
+	cmd.Flags().StringVar(&c.labelFileName, "label-file", "", "Path from repository root to file containing"+
```
Most places we just default this to whatever we do in the core kube repo...

foxish replied:
Done.

eparis commented:
Do you want this enabled in the core kube repo? If so you need to update the submit-queue/deployment.yaml...

foxish commented:
@eparis @janetkuo PTAL

janetkuo commented on mungegithub/features/repo-updates.go:
```diff
@@ -204,6 +206,21 @@ func (o *RepoInfo) LeafAssignees(path string) sets.String {
+// ReadFile returns the contents of a file in the checked out copy of the
+// repository
+func (o *RepoInfo) ReadFile(path string) (string, error) {
+	file, err := os.Open(filepath.Join(o.kubernetesDir, path))
```
Maybe I missed something, but why not just use `ioutil.ReadFile`?

foxish replied:
Done.

foxish commented:
@lavalamp, @apelisse If this looks ok, I'd like to get this merged. We'll need it in the near future and the change just adds the munger but doesn't turn it on.

lavalamp commented on mungegithub/features/repo-updates.go:
```diff
@@ -236,6 +237,17 @@ func (o *RepoInfo) LeafAssignees(path string) sets.String {
 
+// ReadFile returns the contents of a file in the checked out copy of the
+// repository
+func (o *RepoInfo) ReadFile(path string) (string, error) {
```
Generally it's best to return []byte rather than string.

lavalamp commented on mungegithub/features/repo-updates.go:
```diff
@@ -236,6 +237,17 @@ func (o *RepoInfo) LeafAssignees(path string) sets.String {
+// ReadFile returns the contents of a file in the checked out copy of the
+// repository
+func (o *RepoInfo) ReadFile(path string) (string, error) {
+	bytes, err := ioutil.ReadFile(path)
```
Is this entire function useful? I was expecting it to modify the path. Why wouldn't people just call ioutil.ReadFile if they want to do this?

foxish replied:
There used to be path modification before, but not anymore. Removed the function to use ioutil.ReadFile directly.

lavalamp commented on mungegithub/features/repo-updates.go:
```diff
@@ -236,6 +237,17 @@ func (o *RepoInfo) LeafAssignees(path string) sets.String {
+func (o *RepoInfo) ReadFile(path string) (string, error) {
+	bytes, err := ioutil.ReadFile(path)
+	if err != nil {
+		glog.Errorf("%v", err)
```
I recommend not logging if you pass the error up the stack.

lavalamp commented on mungegithub/mungers/check-labels.go:
```diff
@@ -0,0 +1,160 @@
+	return nil
+}
+
+func (c *CheckLabelsMunger) getSha(fileContents string) string {
```
SHA

lavalamp replied:
Actually I'd call it getHash

lavalamp commented on mungegithub/mungers/check-labels.go:
```diff
@@ -0,0 +1,160 @@
+
+func (c *CheckLabelsMunger) getSha(fileContents string) string {
+	h := sha1.New()
+	h.Write([]byte(fileContents))
```
See, here you want a []byte :)

lavalamp replied:
Consider canonicalizing the file before you hash it, so you don't trigger on e.g. sorts or whitespace changes

lavalamp commented on mungegithub/mungers/check-labels.go:
```diff
@@ -0,0 +1,160 @@
+// are created.
+type CheckLabelsMunger struct {
+	labelFilePath string
+	prevSha       string
```
prevHash

lavalamp commented on mungegithub/mungers/check-labels.go:
```diff
@@ -0,0 +1,160 @@
+	prevSha       string
+	labelAccessor labelAccessor
+	features      *features.Features
+	reader        func() (string, error)
```
readFunc

lavalamp commented on mungegithub/mungers/check-labels.go:
```diff
@@ -0,0 +1,160 @@
+	}
+	c.labelAccessor = config
+	c.features = features
+	c.reader = func() (string, error) {
```
Just `c.reader = c.features.Repos.ReadFile` should work

lavalamp commented on mungegithub/mungers/check-labels.go:
```diff
@@ -0,0 +1,160 @@
+	return nil
+}
+
+func (c *CheckLabelsMunger) createNewLabels(repoLabels, fileLabels []*github.Label) {
```
document that this won't remove existing labels.

lavalamp replied:
maybe "addMissingLabels"?

lavalamp commented on mungegithub/mungers/check-labels.go:
```diff
@@ -0,0 +1,165 @@
+func (c *CheckLabelsMunger) Munge(obj *githubhelper.MungeObject) {}
+
+// This function can be used to write a check-labels config file.
+func writeLabelFile(c *CheckLabelsMunger, filename string) error {
```
does anything actually call this function? I recommend omitting it if not.

foxish replied:
Nothing calls it, but it is the way to write out valid labels-config yaml files when building the list from an existing repository. I thought it might be useful to include it.

lavalamp replied:
Dead code rots quickly :)

Add it in another PR if you think it's important.

On Mon, Aug 1, 2016 at 12:05 PM, Anirudh Ramanathan <
notifications@github.com> wrote:

> In mungegithub/mungers/check-labels.go
> <https://github.com/kubernetes/contrib/pull/1306#discussion_r73034291>:
>
> > +			}
> > +		}
> > +	}
> > +}
> > +
> > +// AddFlags will add any request flags to the cobra `cmd`.
> > +func (c *CheckLabelsMunger) AddFlags(cmd *cobra.Command, config *githubhelper.Config) {
> > +	cmd.Flags().StringVar(&c.labelFilePath, "label-file", "", "Path from repository root to file containing"+
> > +		" list of labels")
> > +}
> > +
> > +// Munge is unused by this munger.
> > +func (c *CheckLabelsMunger) Munge(obj *githubhelper.MungeObject) {}
> > +
> > +// This function can be used to write a check-labels config file.
> > +func writeLabelFile(c *CheckLabelsMunger, filename string) error {
>
> Nothing calls it, but it is the way to write out valid labels-config yaml
> files when building the list from an existing repository. I thought it
> might be useful to include it.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/contrib/pull/1306/files/ef4d3000f4599db7762a6d4cef2dd8acbf07c36b#r73034291>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAnglrIBSMcRbehYOxY61clWRajWJoUhks5qbkNqgaJpZM4JDbTs>
> .
>

lavalamp commented on mungegithub/mungers/check-labels_test.go:
```diff
@@ -0,0 +1,128 @@
+
+var (
+	_                              = fmt.Printf
+	_                              = glog.Errorf
```
There shouldn't be a need to actually check in these two lines.

lavalamp commented on mungegithub/mungers/check-labels_test.go:
```diff
@@ -0,0 +1,128 @@
+var (
+	_                              = fmt.Printf
+	_                              = glog.Errorf
+	repoLabelNames []string        = []string{"team/abc", "team/def", "team/ghi", "label/rep"}
```
please just make []*github.Label variables rather than parallel []strings.

foxish replied:
I had trouble setting it directly because the `github.Label` expects `*strings`. I couldn't find a way to create and initialize a `*string` in a single step. Is there a way to do that? Else I could make it a single array of pairs of strings rather than two separate arrays perhaps.

eparis replied:
The tests use `stringPtr()`

foxish replied:
I should have looked harder. Works like a charm. Thanks @eparis.

foxish commented:
@lavalamp, addressed comments, except for canonicalizing the file, because the cost of a re-rerun in case it detects a change in case of spaces/sort order (which themselves seem rare) costs two API calls only. Do you think we should still do it?

lavalamp commented:
LGTM
